### ISSUE TITLE
Trait with unsized fn param compiles without feature gate

### ISSUE THREAD

**Issue by H4x5** [rust-lang/rust#97429]
I created an unimplementable trait with an unsized fn param. I expected an unsized fn param in a trait function to not compile as it is impossible to implement on stable.

I tried this code:

```rust
trait Unimplementable {
    fn foo(x: str);
}

impl Unimplementable for () {
    fn foo(x: str) {}
}
```
<https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=af9c3894dfaa7e09b5fae5b3c83ebb4b>

The trait compiles successfully, but cannot be implemented without the `unsized_fn_params` feature.

Also, the compiler suggests making the param on the impl fn a reference, an invalid signature change.

I expected to see this happen: A compile error on the trait function definition, and the compiler not suggesting an invalid change.

Instead, this happened: The trait compiles, but is impossible to implement without `unsized_fn_params`

Since the trait definition contains an unsized fn param, I would expect such a definition to be feature gated. Unsized fn params in trait definitions should be under the feature gate.

### Meta

`rustc --version --verbose`:
```
rustc 1.62.0-beta.2 (daf68b1f7 2022-05-19)
binary: rustc
commit-hash: daf68b1f766e67ffe040260b15c218301853386a
commit-date: 2022-05-19
host: x86_64-apple-darwin
release: 1.62.0-beta.2
LLVM version: 14.0.4
```

**Comment by lcnr** [comment]
it looks like we generally don't check function headers for well formedness?
```rust
extern {
    fn foo(x: Vec<[str]>); // This compiles
}
trait Unimplemented {
    fn foo(x: Vec<[str]>); // This also compiles
}
```

**Comment by lcnr** [comment]
Ah, so we started to check function arguments for well formedness in Rust 1.4 but have never checked them for being `Sized`. While looking at this, I went to version 1.0 with `x: str` and then only tested `x: Vec<[str]>` using the 1.0 compiler :sweat_smile:

**Comment by compiler-errors** [comment]
So technically the same thing applies for a plain `self` receiver. That is:
```
trait Foo /* :?Sized is implied */ {
  fn foo(self);
}
```
That's an unsized fn param, for example. Similarly you can't implement that trait for anything `?Sized`. 

I wonder if we should have an implied `: Sized` supertrait if we see a receiver that requires `Sized`.
